### PR TITLE
EOS spaces implementation

### PR DESCRIPTION
**Pull request by ishank011**

### CODE REVIEW

**Comment by ishank011** [comment]
@butonic can we merge this? The code smells are mostly from the existing codebase

**Review by glpatcern — APPROVED** [review]

File: internal/grpc/services/storageprovider/storageprovider.go [hunk 1]
```diff
@@ -867,6 +891,12 @@ func (s *service) ListRecycleStream(req *provider.ListRecycleStreamRequest, ss p
 
 	// TODO(labkode): CRITICAL: fill recycle info with storage provider.
 	for _, item := range items {
+		if item.Ref != nil && item.Ref.ResourceId != nil {
+			// The storage driver might set the mount ID by itself, in which case skip this step
+			if mountID, _ := storagespace.SplitStorageID(item.Ref.GetResourceId().GetStorageId()); mountID == "" {
+				item.Ref.ResourceId.StorageId = storagespace.FormatStorageID(s.conf.MountID, item.Ref.GetResourceId().GetStorageId())
+			}
+		}
```
Comment (butonic) [inline internal/grpc/services/storageprovider/storageprovider.go:899]: maybe extract this as `addMissingStorageProviderID` and reuse it everywhere in this file?

File: pkg/storage/registry/spaces/spaces.go [hunk 2]
```diff
@@ -611,9 +626,11 @@ func setPath(space *providerpb.StorageSpace, path string) {
...
 		space.Opaque.Map = map[string]*typesv1beta1.OpaqueEntry{}
 	}
-	space.Opaque.Map["path"] = &typesv1beta1.OpaqueEntry{
-		Decoder: "plain",
-		Value:   []byte(path),
+	if _, ok := space.Opaque.Map["path"]; !ok {
+		space.Opaque.Map["path"] = &typesv1beta1.OpaqueEntry{
+			Decoder: "plain",
+			Value:   []byte(path),
+		}
```
Comment (butonic) [inline pkg/storage/registry/spaces/spaces.go:633]:
```suggestion
		space.Opaque = utils.AppendPlainToOpaque(space.Opaque, "path", path)
```

**Comment by butonic** [comment]
I triggered a ci run in ocis with this branch: https://github.com/owncloud/ocis/pull/3922 it covers some more web cases, eg. for archives.

let's see heow it goes

File: pkg/storage/registry/spaces/spaces.go [hunk 3]
```diff
@@ -501,14 +508,22 @@ func (r *registry) findProvidersForAbsolutePathReference(ctx context.Context, pa
...
 		// when listing paths also return mountpoints
-		filters = append(filters, &providerpb.ListStorageSpacesRequest_Filter{
-			Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_SPACE_TYPE,
-			Term: &providerpb.ListStorageSpacesRequest_Filter_SpaceType{
-				SpaceType: "+mountpoint",
+		filters := []*providerpb.ListStorageSpacesRequest_Filter{
+			{
+				Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_PATH,
+				Term: &providerpb.ListStorageSpacesRequest_Filter_Path{
+					Path: path,
```
[the remark this answers is not in the document]
Comment (ishank011) [reply pkg/storage/registry/spaces/spaces.go]: We'll leave provider paths empty in the registry for now, can be part of a separate PR, let's get this merged

**Comment by ishank011** [comment]
@butonic all tests passed in the ocis repo. Can we merge?

File: internal/grpc/services/gateway/gateway.go [hunk 4]
```diff
@@ -63,15 +63,16 @@ type config struct {
...
+	ShareFolder                  string                            `mapstructure:"share_folder"`
+	DataTransfersFolder          string                            `mapstructure:"data_transfers_folder"`
+	HomeMapping                  string                            `mapstructure:"home_mapping"`
+	TokenManagers                map[string]map[string]interface{} `mapstructure:"token_managers"`
+	EtagCacheTTL                 int                               `mapstructure:"etag_cache_ttl"`
+	AllowedUserAgents            map[string][]string               `mapstructure:"allowed_user_agents"` // map[path][]user-agent
+	CreateHomeCacheTTL           int                               `mapstructure:"create_home_cache_ttl"`
+	ProviderCacheTTL             int                               `mapstructure:"provider_cache_ttl"`
+	StatCacheTTL                 int                               `mapstructure:"stat_cache_ttl"`
+	UseCommonSpaceRootShareLogic bool                              `mapstructure:"use_common_space_root_share_logic"`
```
Comment (butonic) [inline internal/grpc/services/gateway/gateway.go:75]: @ishank011 what is the intention of this flag? does ocis need to set it to true?

Comment (butonic) [reply internal/grpc/services/gateway/gateway.go:75]: ok, got it. CERN will set it to true to store spacegrants in the db as well.

File: pkg/storage/registry/spaces/spaces.go [hunk 5]
```diff
@@ -501,14 +509,22 @@ func (r *registry) findProvidersForAbsolutePathReference(ctx context.Context, pa
...
 		// when listing paths also return mountpoints
-		filters = append(filters, &providerpb.ListStorageSpacesRequest_Filter{
-			Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_SPACE_TYPE,
-			Term: &providerpb.ListStorageSpacesRequest_Filter_SpaceType{
-				SpaceType: "+mountpoint",
+		filters := []*providerpb.ListStorageSpacesRequest_Filter{
+			{
+				Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_PATH,
+				Term: &providerpb.ListStorageSpacesRequest_Filter_Path{
+					Path: strings.TrimSuffix(path, p.ProviderPath),
```
Comment (butonic) [inline pkg/storage/registry/spaces/spaces.go:518]: oops
```suggestion
					Path: strings.TrimPrefix(path, p.ProviderPath),
```

**Review by butonic — APPROVED** [review]

**Pull request merged**